### ISSUE TITLE
Crash on key press

### ISSUE THREAD

**Issue by beetlenaut** [DFHack/dfhack#1837]
Game crashes to the desktop randomly after a few minutes, making it unplayable. Crashes almost always seem to be triggered by a key being pressed. It has happened more than once with F1 and with Z-down. I didn't have this issue with dfhack for 0.47.04 on the same computer. The vanilla game also seems to be stable (though I much prefer graphics).

Steps to reproduce: Launch Dwarf Fortress with dfhack and play for 10-15 minutes.

dfhack: 0.47.05-r1
OS: openSUSE Linux Leap 15.2 (latest release)
Tileset: Meph
Happens with or without PyLNP. Happens with TWBT or TWBT_LEGACY selected in PyLNP.

Tail of strace output:
>8833  exit_group(0)                     = ?
8833  +++ exited with 0 +++
8342  <... wait4 resumed>[{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 8833
8342  rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f0939e174a0}, {sa_handler=0x5570a8d40d10, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f0939e174a0}, 8) = 0
8342  ioctl(2, TIOCGWINSZ, {ws_row=69, ws_col=115, ws_xpixel=0, ws_ypixel=0}) = 0
8342  rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
8342  --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=8833, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
8342  wait4(-1, 0x7ffcebef0e90, WNOHANG, NULL) = -1 ECHILD (No child processes)
8342  rt_sigreturn({mask=[]})           = 0
8342  read(255, "echo\n\nif [ -n \"$DF_POST_CMD\" ]; "..., 5694) = 73
8342  fstat(1, {st_mode=S_IFCHR|0600, st_rdev=makedev(0x88, 0x3), ...}) = 0
8342  write(1, "\n", 1)                 = 1
8342  rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
8342  rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
8342  exit_group(134)                   = ?
8342  +++ exited with 134 +++

Let me know what else would help, or if there is something I could try that might reproduce it more reliably.

**Comment by lethosor** [comment]
Could you run with GDB with `./dfhack -g` (followed by `r`) and include the full output of `bt` when the crash occurs?

Also, could you try disabling some things, roughly in this order, to see if they stop the crash?
- `disable search`
- `tweak stable-cursor disable`
- `unload -all`

Output from `install-info` would also be helpful.

**Comment by beetlenaut** [comment]
It does seem to be a problem with TwbT. The end of the backtrace indicates that, and I realized that the only keys that I have seen cause a crash would change some text on the screen.

Using 2D with TTF doesn't cause crashes, but it's not really a solution. It's too hard to play because important text isn't legible, and half the tiles are wrong. There have been a lot of improvements since that was current!

[bt.txt](https://github.com/DFHack/dfhack/files/6322659/bt.txt) -- I hope this is the output you wanted.
[install-info.txt](https://github.com/DFHack/dfhack/files/6322658/install-info.txt)

**Comment by lethosor** [comment]
Thanks! The only thing not quite clear to me from the `bt` output is which thread the crash originated in. Thread 1 indicates a TWBT call, and that would line up with your other observations as well.

Unfortunately TWBT isn't a plugin we officially support, at least not in this repo - it's currently maintained by @thurin at https://github.com/thurin/df-twbt. From a glance at the [changes since 0.47.04](https://github.com/thurin/df-twbt/compare/1463d16b3642c2c5b88cb5b7d909a8f3b3f70350...21b6c4860fc74e1295445566fe954cab816323d0), I'm not seeing an obvious reason why you would be getting a _new_ crash in 0.47.05, unless you were using DFHack 0.47.04-r1 or older previously. Do you know what DFHack version you were using before?

The backtrace indicates a call to `abort()`, which sometimes prints additional information to standard error. Is there anything printed to the terminal when this crash occurs, or is anything added at the end of `stderr.log`?

Also, there are several lines like this in the backtrace where it would be useful to get more information:
```
#2  0x00007ffff64ce016 in  () at /usr/lib64/libstdc++.so.6
#3  0x00007ffff64d988c in  () at /usr/lib64/libstdc++.so.6
#4  0x00007ffff64d98f7 in  () at /usr/lib64/libstdc++.so.6
#5  0x00007ffff64d9ba9 in  () at /usr/lib64/libstdc++.so.6
```
Does GDB give you a "Missing separate debuginfos" message? If so, could you try installing the packages it specifies? [This thread](https://www.reddit.com/r/openSUSE/comments/6j261v/trying_to_install_debug_symbols/) may be helpful, although I've never tried this on OpenSUSE before.

**Comment by beetlenaut** [comment]
I can't add an issue to thurin/df-twbt, so I don't know what I would need to *do* now.

I *was* using 0.47.04-r1 before. I thought that was the previous one. I didn't have any crashes with that, but I wanted to try out the new version of DF: It sounded like a lot of issues were fixed. The first DFHack version that caused these crashes was 0.47.05-beta1.

The only message I got in the terminal was "Abort (core dumped)." There is nothing about a crash in `stderr.log`; just a lot of "Invoking" and "Loading."

I had literally 100's of messages about "Missing separate debuginfos." It went on for pages, so there is no way I can install them all. (I'm pretty much out of time for a couple months, so I'm not going to try writing a script or something.)

**Comment by lethosor** [comment]
> I _was_ using 0.47.04-r1 before. I thought that was the previous one.

Just to clarify, was this the *pack* version or the *DFHack* version? DFHack went all the way up to 0.47.04-**r5** for DF 0.47.04, but I know some packs also use the "r#" versioning scheme, and I don't know for sure about yours. If you still have that installation, running `help` in the DFHack console would give you the DFHack version number.

Incidentally, could you provide the link to the pack you're using? PyLNP is just the launcher, and there are multiple packs for Linux that use it.

> I had literally 100's of messages about "Missing separate debuginfos." It went on for pages, so there is no way I can install them all.

If there's one that mentions "libstdc++", I think that would be the only one necessary to get a more complete stack trace in this case. Failing that, anything "gcc"-related might work too.

**Comment by beetlenaut** [comment]
> Just to clarify, was this the pack version or the DFHack version?

Apparently just the pack. Running 'help' on the old installation gives me 'DFHack version 0.47.04-alpha0'.

I got the pack [here](https://dffd.bay12games.com/file.php?id=15054). I didn't know there were other Linux packs, but I would like to know where *they* can be found for future reference.

In other news, I saw something about a possible Meph tileset bug at [github.com/thurin/df-twbt](https://github.com/thurin/df-twbt). I decided I had time to try one more time using a different tileset (Vettlingr). That fixed it! So, this seems to be a TwbT problem specific to Meph, and it has a simple workaround. If you still want a complete stack trace, I can work on that some other time.
